src/toil_lib/test/test_programs.py

```diff
@@ -1,14 +1,62 @@
-    docker_call(work_dir=work_dir, parameters=parameter, tool=tool)
-    # Test outfile
-    fpath = os.path.join(work_dir, 'test')
-    with open(fpath, 'w') as f:
-        docker_call(tool='ubuntu', env=dict(foo='bar'), parameters=['printenv', 'foo'], outfile=f)
-    assert open(fpath).read() == 'bar\n'
+def test_docker_clean(tmpdir):
+    """
+    Run the test container that creates a file in the work dir, and sleeps for 5 minutes.  Ensure
+    that the calling job gets SIGKILLed after a minute, leaving behing the spooky/ghost/zombie
+    container. Ensure that the container is killed on batch system shutdown (through the defer
+    mechanism).
+
+    This inherently also tests docker_call
+
+    :param tmpdir: A temporary directory to work in
+    :returns: None
+    """
+    data_dir = os.path.join(str(tmpdir), 'data')
+    work_dir = os.path.join(str(tmpdir), 'working')
+    os.mkdir(data_dir)
+    os.mkdir(work_dir)
+    options = Job.Runner.getDefaultOptions(os.path.join(str(tmpdir), 'jobstore'))
+    options.logLevel = 'INFO'
+    options.workDir = work_dir
+    A = Job.wrapJobFn(_test_docker_clean_fn, data_dir)
+    try:
+        Job.Runner.startToil(A, options)
+    except FailedJobsException:
+        # The file created by spooky_container would remain in the directory.  Since it was created
+        # inside the container, it would have had uid and gid == 0 (root) upon creation.  If the
+        # defer mechanism worked well, it should now be non-zero and we check for that.
+        file_stats = os.stat(os.path.join(data_dir, 'test.txt'))
+        assert file_stats.st_gid != 0
+        assert file_stats.st_uid != 0
+        # Ensure the container isn't still running.
+        with open(os.path.join(data_dir, 'container.name'), 'r') as file_handle:
+            container_name = file_handle.read()
+        container_prefix = '--'.join(container_name.split('--')[:2])
+        running_containers = subprocess.check_output(['docker', 'ps', '-a', '--format',
+                                                      '"{{.Names}}"'])
+        assert container_prefix not in running_containers
+
+
+def _test_docker_clean_fn(job, data_dir):
+    def kill_self():
+        test_file = os.path.join(data_dir, 'test.txt')
+        # This will kill the worker once we are sure the docker container started
+        while not os.path.exists(test_file):
+            time.sleep(1)
+        # By the time we reach here, we are sure the cotainer is running.  Write down the container
+        # prefix so we can check for it later.
+        with open(os.path.join(data_dir, 'container.name'), 'w') as file_handle:
+            file_handle.write(_get_container_name(job))
+        os.kill(os.getpid(), signal.SIGKILL)  # signal.SIGINT)
+        # TODO: FIX OCCASIONAL DEADLOCK WITH SIGINT (tested on single machine)
+    Thread(target=kill_self).start()
+    docker_call(job, tool='quay.io/ucsc_cgl/spooky_test', work_dir=data_dir, daemon=False)
```

Comment:
where? The immediate next call is docker_call and we are guaranteed that it will never finish